File: src/schemas/agentSchemas.ts
import { z } from 'zod'

// 1. Home Agent Action Schema
// Used in HomePage.vue when useAgentMode is true.
export const HomeAgentActionSchema = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('insert_text'),
    agent_reasoning: z.string().describe('Explanation for the Activity Feed'),
    content: z.string(),
    location: z.enum(['cursor', 'replace_selection', 'end']),
  }),
  z.object({
    type: z.literal('modify_selection'),
    agent_reasoning: z.string().describe('Explanation for the Activity Feed'),
    instructions: z.string(),
  }),
  z.object({
    type: z.literal('execute_tool'),
    agent_reasoning: z.string().describe('Explanation for the Activity Feed'),
    tool_name: z.string().describe('Must be an enabled WordToolName or GeneralToolName'),
    arguments: z.record(z.any()),
  }),
  z.object({
    type: z.literal('request_user_clarification'),
    agent_reasoning: z.string(),
    question: z.string(),
  }),
  z.object({
    type: z.literal('no_action'),
    agent_reasoning: z.string(),
    reason: z.string(),
  }),
])

export type HomeAgentAction = z.infer<typeof HomeAgentActionSchema>

// 2. Advisor Workflow Action Schema
// Used in ConsultantPage.vue -> Advisor Sub-mode.
export const AdvisorActionSchema = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('highlight_critical_range'),
    agent_reasoning: z.string().describe('Why this range is critical'),
    text_to_highlight: z.string(),
    reason: z.string().describe('User-facing reason displayed in UI'),
    severity: z.enum(['low', 'medium', 'high']),
  }),
  z.object({
    type: z.literal('propose_document_edit'),
    agent_reasoning: z.string(),
    original_text: z.string(),
    proposed_text: z.string(),
    explanation: z.string().describe('Why the edit is proposed'),
  }),
  z.object({
    type: z.literal('request_approval'),
    agent_reasoning: z.string(),
    summary: z.string(),
    choices: z.array(z.string()).optional().describe("E.g., ['Approve', 'Reject', 'Modify']"),
  }),
  z.object({
    type: z.literal('proceed_to_next_step'),
    agent_reasoning: z.string(),
    current_step_index: z.number(),
    step_summary: z.string(),
  }),
])

export type AdvisorAction = z.infer<typeof AdvisorActionSchema>

// 3. Create Document Setup Schema
// Used in ConsultantPage.vue -> Create Document Sub-mode.
export const CreateDocumentSetupSchema = z.object({
  agent_reasoning: z.string().describe("Agent's thought process behind this document structure"),
  title: z.string().describe('Primary title of the generated document'),
  description: z.string().describe('Short abstract or executive summary'),
  sections: z.array(
    z.object({
      title: z.string(),
      content: z.string(),
      requires_source_verification: z.boolean().default(false),
      source_reference: z.string().optional().describe('Name of the DMS file or source of truth used'),
    }),
  ),
  metadata: z.object({
    industry: z.string(),
    documentType: z.string(),
    targetAudience: z.string().optional(),
    tone: z.enum(['professional', 'academic', 'casual', 'persuasive', 'direct']).optional(),
  }),
})

export type CreateDocumentSetup = z.infer<typeof CreateDocumentSetupSchema>

File: backend/agent_schemas.py
from typing import Any, Dict, List, Literal, Optional, Union
from pydantic import BaseModel, Field

# --- Base Schemas ---

class AgentActionBase(BaseModel):
    agent_reasoning: str = Field(..., description="Explanation for the Activity Feed or internal log")

# --- Home Agent Actions ---

class InsertTextAction(AgentActionBase):
    type: Literal["insert_text"] = "insert_text"
    content: str
    location: Literal["cursor", "replace_selection", "end"]

class ModifySelectionAction(AgentActionBase):
    type: Literal["modify_selection"] = "modify_selection"
    instructions: str

class ExecuteToolAction(AgentActionBase):
    type: Literal["execute_tool"] = "execute_tool"
    tool_name: str = Field(..., description="Must be an enabled WordToolName or GeneralToolName")
    arguments: Dict[str, Any]

class RequestClarificationAction(AgentActionBase):
    type: Literal["request_user_clarification"] = "request_user_clarification"
    question: str

class NoAction(AgentActionBase):
    type: Literal["no_action"] = "no_action"
    reason: str

HomeAgentAction = Union[
    InsertTextAction,
    ModifySelectionAction,
    ExecuteToolAction,
    RequestClarificationAction,
    NoAction
]

# --- Advisor Actions ---

class HighlightRangeAction(AgentActionBase):
    type: Literal["highlight_critical_range"] = "highlight_critical_range"
    text_to_highlight: str
    reason: str = Field(..., description="User-facing reason displayed in UI")
    severity: Literal["low", "medium", "high"]

class ProposeEditAction(AgentActionBase):
    type: Literal["propose_document_edit"] = "propose_document_edit"
    original_text: str
    proposed_text: str
    explanation: str = Field(..., description="Why the edit is proposed")

class RequestApprovalAction(AgentActionBase):
    type: Literal["request_approval"] = "request_approval"
    summary: str
    choices: Optional[List[str]] = Field(None, description="E.g., ['Approve', 'Reject', 'Modify']")

class ProceedStepAction(AgentActionBase):
    type: Literal["proceed_to_next_step"] = "proceed_to_next_step"
    current_step_index: int
    step_summary: str

AdvisorAction = Union[
    HighlightRangeAction,
    ProposeEditAction,
    RequestApprovalAction,
    ProceedStepAction,
    ExecuteToolAction
]

# --- Create Document Action ---

class DocumentSection(BaseModel):
    title: str
    content: str
    requires_source_verification: bool = False
    source_reference: Optional[str] = Field(None, description="Name of the DMS file or source of truth used")

class CreateDocumentSetup(AgentActionBase):
    title: str = Field(..., description="Primary title of the generated document")
    description: str = Field(..., description="Short abstract or executive summary")
    sections: List[DocumentSection]
    metadata: Dict[str, Any] = Field(..., description="Industry, documentType, tone, etc.")
